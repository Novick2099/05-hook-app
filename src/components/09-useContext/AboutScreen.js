import React from 'react'

export const AboutScreen = () => {
    return (
        <div>
            <h2>AboutScreen</h2>
            <hr/>
        </div>
    )
}

