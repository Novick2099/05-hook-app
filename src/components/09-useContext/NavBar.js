import React from 'react'
import { Link, NavLink } from 'react-router-dom'
import ResponsiveDrawer from './ResponsiveDrawer'

export const NavBar = () => {
    return (
        <div>
            <ul>
                {/* todo NavLink a diferencia de Link tiene la propiedad activeClassName="" para aplicar cuando estemos en esa ruta,
                 para que funcione la clase, NavLink debe llevar la palabra exact */}
                <li><NavLink exact activeClassName='clase' to="/">Home</NavLink></li>
                <li> <NavLink exact to="/about">About</NavLink> </li>
                <li> <NavLink exact to="/login">Login</NavLink></li>
            </ul>
        </div>
    )
}
