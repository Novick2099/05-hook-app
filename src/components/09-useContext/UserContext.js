import { createContext } from 'react';

//createContext Higher Order Component
export const UserContext = createContext(null);

