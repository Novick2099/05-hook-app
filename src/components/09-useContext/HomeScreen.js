import React, {useContext} from 'react'
import {UserContext} from "./UserContext";



export const HomeScreen = () => {

//todo renombrar en desestructuracion
// todo Buscar la instancia dentro del arbol de componentes que sea de tipo UserContext
  const { user: usuario } = useContext( UserContext );

  console.log(usuario);

  return (
        <div>
            <h1>Home Screen</h1>
            <hr/>
          <p>
            {JSON.stringify( usuario, null, 3 )}
          </p>
        </div>
    )
}
