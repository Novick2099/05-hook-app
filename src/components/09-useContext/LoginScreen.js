import React, {useContext} from 'react'
import {UserContext} from "./UserContext";
import {Button} from "@mui/material";


export const LoginScreen = () => {

  // 1 obtener la referencia al context
  // 2 set User
  // crear un boton para llamar la funcion set User

  const testUser = {
    id: 123,
    name: 'Carlos',
    email: 'carlos@gmail.com'
  }

  const {user: userInContext, setUser: setUserInContext} = useContext(UserContext);



  const handleSetUser = ()=> {
    setUserInContext(testUser)
  }

    return (
        <div>
            <h1>Login Screen</h1>
            <hr/>
          <Button id={"set_user"} onClick={handleSetUser} variant="outlined" color="primary">
            Set User
          </Button>
        </div>
    )
}
