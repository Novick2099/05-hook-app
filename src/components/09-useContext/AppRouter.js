import React from "react";
import { BrowserRouter as Router, Switch, Route, Link, Redirect } from "react-router-dom";
import { AboutScreen } from "./AboutScreen";
import { HomeScreen } from "./HomeScreen";
import { LoginScreen } from "./LoginScreen";
import { NavBar } from "./NavBar";

export const AppRouter = () => {
  return (
    <Router>
      <div>
        <NavBar />
        {/* todo A <Switch> looks throught its children <Route>s and renders
        the first one that matches the current URL */}
        <Switch>
          {/* todo exact para forzar a que la url sea exacta*/}
          <Route exact path="/" component={HomeScreen} />
          <Route path="/about" component={AboutScreen} />
          <Route path="/login" component={LoginScreen} />

          {/*Cuando la ruta no se encuentra, se redirecciona hacia lo especificado por el redirect*/}
          <Redirect to="/" />
        </Switch>
      </div>
    </Router>
  );
};
