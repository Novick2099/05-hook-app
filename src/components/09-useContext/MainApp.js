import React, {useState} from 'react'
import { AppRouter } from './AppRouter'
import {Typography} from "@mui/material";
import {UserContext} from "./UserContext";

export const MainApp = () => {

  const [user, setUser] = useState({});

  const userInitialState = {
    id: 1234,
    name: 'Carlos',
    email: 'carlos@gmail.com',
  }
    return (
      /*Se guarda en el contexto usando esta notacion,
      lo que este siendo envuelto por el userContext compartira estado*/
      /*
      {user, setUser} es la forma corta de {user:user, setUser: setUser}
      * */
      <UserContext.Provider value={{
        user,
        setUser,
      } }>
        <Typography variant={"h1"}>Main App No debe cambiar</Typography>
        <AppRouter/>
      </UserContext.Provider>
    )
}
