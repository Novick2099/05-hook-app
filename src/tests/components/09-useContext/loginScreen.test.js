import React from 'react';
import {mount} from "enzyme";
import {UserContext} from "../../../components/09-useContext/UserContext";
import {LoginScreen} from "../../../components/09-useContext/LoginScreen";
import {Button} from "@mui/material";


describe('',()=>{

  const user ={
    name: 'Carlos',
    email: 'carlos@gmail.com'
  }

  const setUser = jest.fn();

  const wrapper = mount(
    <UserContext.Provider value={{
      setUser
    }}>
      <LoginScreen />
    </UserContext.Provider>
  );

  it('Should Match Snapshot',()=>{
    expect(wrapper).toMatchSnapshot();
  });

  it('Should simulate with expected argument',()=>{
    //todo llamar a una funcion en un test con parametros
    //wrapper.find('button').prop('onClick')();
    wrapper.find('#set_user').last().simulate('click');
    expect(setUser).toHaveBeenLastCalledWith({
      id: 123,
      name: 'Carlos',
      email: 'carlos@gmail.com'
    });
  });


});