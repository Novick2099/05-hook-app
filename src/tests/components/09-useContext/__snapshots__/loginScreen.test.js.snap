// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` Should Match Snapshot 1`] = `
<div>
  <h1>
    Login Screen
  </h1>
  <hr />
  Array [
    "",
    Array [
      "",
      <button
        className="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        Set User
        Array [
          "",
          <span
            className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />,
        ]
      </button>,
    ],
  ]
</div>
`;
